File: packages/flet/lib/src/controls/semantics_service.dart
import 'package:flutter/material.dart';
import 'package:flutter/semantics.dart';

import '../flet_control_backend.dart';
import '../models/control.dart';
import '../utils/others.dart';

class SemanticsServiceControl extends StatelessWidget {
  final Control? parent;
  final Control control;
  final FletControlBackend backend;

  const SemanticsServiceControl(
      {super.key, this.parent, required this.control, required this.backend});

  @override
  Widget build(BuildContext context) {
    debugPrint("SemanticsService build: ${control.id}");

    backend.subscribeMethods(control.id, (methodName, args) async {
      var message = args["message"].toString();
      switch (methodName) {
        case "announce_message":
          debugPrint("SemanticsService.announceMessage($message)");
          var rtl = args["rtl"] == "true";
          var assertiveness = parseAssertiveness(
              args["assertiveness"].toString(), Assertiveness.polite)!;
          SemanticsService.announce(
              message, rtl ? TextDirection.rtl : TextDirection.ltr,
              assertiveness: assertiveness);
          break;
        case "announce_tooltip":
          debugPrint("SemanticsService.announceTooltip($message)");
          SemanticsService.tooltip(message);
          break;
      }
      return null;
    });

    return const SizedBox.shrink();
  }
}

File: sdk/python/packages/flet/src/flet/controls/core/semantics_service.py
import asyncio
from enum import Enum

from flet.controls.control import Service, control

__all__ = ["SemanticsService", "Assertiveness"]


class Assertiveness(Enum):
    POLITE = "polite"
    ASSERTIVE = "assertive"


@control("SemanticsService")
class SemanticsService(Service):
    async def announce_tooltip_async(self, message: str):
        await self._invoke_method_async(
            "announce_tooltip", arguments={"message": message}
        )

    def announce_tooltip(self, message: str):
        asyncio.create_task(self.announce_tooltip_async(message))

    async def announce_message_async(
        self,
        message: str,
        rtl: bool = False,
        assertiveness: Assertiveness = Assertiveness.POLITE,
    ):
        await self._invoke_method_async(
            "announce_message",
            arguments={
                "message": message,
                "rtl": str(rtl),
                "assertiveness": assertiveness,
            },
        )

    def announce_message(
        self,
        message: str,
        rtl: bool = False,
        assertiveness: Assertiveness = Assertiveness.POLITE,
    ):
        asyncio.create_task(self.announce_message_async(message, rtl, assertiveness))
